connections/stream-migration.md

```diff
@@ -0,0 +1,217 @@
+    it will close <b>stream A</b> for writing.
+    It will only write to <b>stream B</b> from now on.
+end note
+
+Initiator <- Responder: <b>Stream A:</b> ""EOF""
+
+note over Initiator
+    Treat any ""EOF"" on <b>stream A</b> as a signal
+    that it should continue reading on <b>stream B</b>
+end note
+
+note over Initiator, Responder
+    At this point <b>stream A</b> is closed for writing on
+    both sides, and both sides have read up to ""EOF"".
+    <b>stream A</b> has been fully migrated to <b>stream B</b>
+end note
+
+@enduml
+```
+
+To generate:
+```bash
+plantuml stream-migration.md -o stream-migration -tsvg
+```
+</details>
+
+Note: some of these steps may be pipelined.
+
+### Resets
+
+If either stream is "reset" before both ends are closed, both streams must be
+reset and the stream as a whole should be considered "aborted" (reset).
+
+### Half closed streams
+
+The final migrated stream should loook the same as the initial stream. If the
+initial stream `A` was half closed, then the final migrated stream `B` should
+also be half closed. Note this may involve an extra step by one of the nodes.
+If a node, when trying to close writes to its old stream, notices that it was
+already closed, it should also close the new stream for writing. Specifically
+imagine the following case.
+
+
+![stream-migration-half-closed](./stream-migration/stream-migration-half-closed.svg)
+
+<details>
+  <summary>Instructions to reproduce diagram</summary>
+``` plantuml
+@startuml stream-migration-half-closed
+skinparam sequenceMessageAlign center
+entity Initiator
+entity Responder
+
+note over Responder: <b>Stream A</b> is closed for writing
+
+== Stream Migration ==
+
+note over Initiator, Responder: Migrate <b>Stream A</b> to <b>Stream B</b>
+
+Initiator -> Responder: <b>Stream A:</b> Start stream migration, this stream is <b>id=A</b>
+
+Initiator -> Responder: Open new stream on <b>Connection 2</b>. Call this <b>Stream B</b>
+
+Initiator -> Responder: <b>Stream B:</b> Migrate stream with <b>id=A</b> to this stream
+Initiator <- Responder: <b>Stream B:</b> Ack Migrate
+
+note over Initiator
+    Close <b>stream A</b> for writing.
+    Will only write to <b>stream B</b> from now on.
+end note
+
+
+note over Initiator
+    We have already seen the ""EOF"" on
+    <b>Stream A</b> from <i>Responder</i>
+    So we continue reading on <b>stream B</b>
+end note
+
+Initiator -> Responder: <b>Stream A:</b> ""EOF""
+
+note over Responder
+    Treat ""EOF"" on <b>stream A</b> as a signal to close <b>stream A</b> for
+    writing and continue writing on <b>stream B</b>. However stream A was
+    already closed (before migration), so we close <b>stream B</b> as well here.
+end note
+Initiator <- Responder: <b>Stream B:</b> ""EOF""
+
+
+note over Initiator, Responder: Stream A is now migrated to Stream B
+@enduml
+```
+To generate:
+```bash
+plantuml stream-migration.md -o stream-migration -tsvg
+```
+</details>
+
```

Comment:
What happens if the peer misbehaves, e.g. when B just doesn’t send an EOF on stream B? While A can “consider” the stream as closed, it still needs to be closed explicitly, otherwise the stream multiplexer can’t garbage-collect the stream after it has been EOFed from both sides.

Reply:
I think if the Responder misbehaves and doesn't close B then the stream migration hasn't finished since the new stream isn't in the same state as the old stream. I'm not sure what else we can do besides consider it not spec-compliant.